ceph: add support for hybrid Storage Pools

sp98 opened the pull request:
Adding support for hybrid Storage pools to store replicas on different storage tiers. This PR creates a hybrid crush rule for choosing Primary OSD for high performing SSD devices and remaining OSD for low performance HDD devices.

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #6805 

References:
https://docs.ceph.com/en/latest/rados/operations/crush-map/#custom-crush-rules

TODO:

- [x] Update CephBlockPool CR in case of invalid device classes
- [x] Test on a cluster with SSD and HDD disks.

Testing:
Tested on minikube cluster by manually adding ssd device class

1. Removed hdd from osd.O 
  ceph osd crush rm-device-class osd.0

2. Add `sdd` device class for OSD.0
 ceph osd crush set-device-class ssd OSD.0

3. 
sh-4.4# ceph osd tree 
```
ID  CLASS  WEIGHT   TYPE NAME              STATUS  REWEIGHT  PRI-AFF
-1         0.02998  root default                                    
-5         0.00999      host minikube-m02                           
 1    hdd  0.00999          osd.1              up   1.00000  1.00000
-3         0.00999      host minikube-m03                           
 0    ssd  0.00999          osd.0              up   1.00000  1.00000
-7         0.00999      host minikube-m04                           
 2    hdd  0.00999          osd.2              up   1.00000  1.00000
```

4. Created hybrid storage pool
sh-4.4# ceph osd crush rule dump hybridpool
```{
    "rule_id": 1,
    "rule_name": "hybridpool",
    "ruleset": 1,
    "type": 1,
    "min_size": 1,
    "max_size": 10,
    "steps": [
        {
            "op": "take",
            "item": -12,
            "item_name": "default~ssd"
        },
        {
            "op": "chooseleaf_firstn",
            "num": 1,
            "type": "host"
        },
        {
            "op": "emit"
        },
        {
            "op": "take",
            "item": -2,
            "item_name": "default~hdd"
        },
        {
            "op": "chooseleaf_firstn",
            "num": 0,
            "type": "host"
        },
        {
            "op": "emit"
        }
    ]
}
```


5. Adding data to the hybrid storage pool
sh-4.4# cat hello.txt 
hello world!
sh-4.4# rados -p hybridpool put object1 hello.txt 
sh-4.4# rados -p hybridpool ls
object1
sh-4.4# ceph -s
 ```
 cluster:
    id:     de1ff337-eca6-43c2-a4ca-d1cd1f5c6e60
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum a,b,c (age 13m)
    mgr: a(active, since 11m)
    osd: 3 osds: 3 up (since 12m), 3 in (since 12m)
 
  data:
    pools:   2 pools, 33 pgs
    objects: 4 objects, 13 B
    usage:   16 MiB used, 30 GiB / 30 GiB avail
    pgs:     33 active+clean
```
 
sh-4.4# ceph osd map hybridpool ojbect1 -f json-pretty    

```
{
    "epoch": 23,
    "pool": "hybridpool",
    "pool_id": 2,
    "objname": "ojbect1",
    "raw_pgid": "2.96261f6f",
    "pgid": "2.f",
    "up": [
        0,
        1,
        2
    ],
    "up_primary": 0,
    "acting": [
        0,
        1,
        2
    ],
    "acting_primary": 0
}
```
sh-4.4# 

6. OSD.0 seems to be the primary osd.

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

sp98 commented:
Question:  When device classes provided in the Hybrid Storage Pool spec are not valid, then do we need to continue with pool creation and show a warning message in the pool spec status, or do we need to abort the creation of the crush rule? I remember we agreed on the former, but existing validation (in `[validatePool](https://github.com/rook/rook/blob/056150484f4792cff7bf14665af83fc16ea7e105/pkg/operator/ceph/pool/validate.go#L29)()` method) returns error if any other spec is not valid?  @travisn

sp98 commented:
> `validatePool()` is intended to validate the spec settings, but not actually query Ceph for anything. We should create the pool without checking for the existence of the device class. It will help remove ordering dependency of resource creation. In other words, the OSDs with the new deviceClass might still be starting. On the pool status it would be nice to show a message that the device class was not found, but I don't see it as critical.

While testing it more, it seems that ceph is not compiling the crush map if the device class is not valid.
```ceph-block-pool-controller: failed to reconcile. failed to create pool "hybridpool".: failed to create pool "hybridpool".: failed to create pool "hybridpool": failed to create hybrid crush rule "hybridpool": failed to compile crush map: failed to compile crush map "/tmp/278633651.decompiled.compiled". .:```

Same result is scene when doing it manually.  Need to try it once on a cluster with mixed device types to confirm this behavior.

leseb commented on pkg/operator/ceph/pool/validate.go:
```diff
@@ -163,3 +165,47 @@ func ValidatePoolSpec(context *clusterd.Context, clusterInfo *cephclient.Cluster
+	}
+
+	if len(invalidDeviceClasses) > 0 {
+		msg := fmt.Sprintf("CephBlockPool %q has invalid device classes: %v", p.Name, invalidDeviceClasses)
```
I don't think this is sufficient, we must check with `ceph osd crush class ls-osd $TYPE`, and each `type` must return at least one element.

sp98 replied:
Thanks for pointing that out. Using `ceph osd crush class ls-osd $TYPE` to check if device class is valid or not.

sp98 commented on pkg/daemon/ceph/client/crush_rule.go:
```diff
@@ -62,6 +76,30 @@ func buildTwoStepPlainCrushRule(crushMap CrushMap, ruleName string, pool cephv1.
+	var primaryOSDDeviceClass string
+	var secondaryOSDsDeviceClass string
+	if pool.Replicated.HybridStorage.PrimaryDeviceClass != "" {
+		primaryOSDDeviceClass = fmt.Sprintf("class %s", pool.Replicated.HybridStorage.PrimaryDeviceClass)
```
moved `class` in the template.

sp98 commented on cluster/examples/kubernetes/ceph/pool.yaml:
```diff
@@ -22,6 +22,9 @@ spec:
     # replicasPerFailureDomain: 2
     # The name of the failure domain to place further down replicas
     # subFailureDomain: host
+    # hybridStorage:
```
moved it above, to line 21

sp98 commented on pkg/operator/ceph/pool/validate.go:
```diff
@@ -163,3 +165,47 @@ func ValidatePoolSpec(context *clusterd.Context, clusterInfo *cephclient.Cluster
+		return nil, nil
+	}
+
+	invalidDeviceClasses := []string{}
```
using `ceph osd crush class ls-osd $TYPE` command now. So this `invalidDeviceClasses` is not used anymore.

leseb commented on Documentation/ceph-pool-crd.md:
````diff
@@ -31,6 +31,26 @@ spec:
+      primaryDeviceClass: ssd
+      secondaryDeviceClass: hdd
+```
+Note: Ensure that `primaryDeviceClass` and `secondaryDeviceClass` are valid and they have at least one OSD associated with them, else the crush rule re-compile will fail.
````
I agree with the warning although I'll remove the `else the crush rule re-compile will fail.
Write`, it feels like an internal details.
```suggestion
Warning: Ensure that `primaryDeviceClass` and `secondaryDeviceClass` are valid and they have at least one OSD associated with them, else the crush rule re-compile will fail.
```

leseb commented on pkg/operator/ceph/pool/validate.go:
```diff
@@ -41,6 +43,14 @@ func ValidatePool(context *clusterd.Context, clusterInfo *cephclient.ClusterInfo
+	if p.IsHybridStoragePool() {
+		err := validateDeviceClasses(context, clusterInfo, p)
+		if err != nil {
+			return errors.Wrapf(err, "failed to validate device classes for hybrid storage pool spec")
```
```suggestion
			return errors.Wrap(err, "failed to validate device classes for hybrid storage pool spec")
```

leseb commented on pkg/operator/ceph/pool/validate.go:
```diff
@@ -163,3 +173,37 @@ func ValidatePoolSpec(context *clusterd.Context, clusterInfo *cephclient.Cluster
+		return errors.Wrapf(err, "failed to get osds for the device class %q", deviceClassName)
+	}
+	if len(deviceClassOSDs) == 0 {
+		return errors.New(fmt.Sprintf("no osds available for the device class %q", deviceClassName))
```
```suggestion
		return errors.New(fmt.Sprintf("no osds available for the device class %q", deviceClassName))
```

use `errors.Errorf()`

leseb commented on pkg/daemon/ceph/client/pool.go:
```diff
@@ -582,3 +615,24 @@ func GetPoolStatistics(context *clusterd.Context, clusterInfo *ClusterInfo, name
 	return &poolStats, nil
 }
+
+func crushRuleExists(crushMap CrushMap, ruleName string) bool {
```
`util.NewSet()`?

sp98 replied:
`crushMap` is obtained by running a ceph command and un-marshalling result  into `CrushMap` struct (in `GetCrushMap` method).  

So don't think I can use `util.NewSet()` here.

leseb commented on Documentation/ceph-pool-crd.md:
````diff
@@ -31,6 +31,26 @@ spec:
+      primaryDeviceClass: ssd
+      secondaryDeviceClass: hdd
+```
+Warning: Ensure that `primaryDeviceClass` and `secondaryDeviceClass` are valid and they have at least one OSD associated with them, else the crush rule re-compile will fail.
````
Somehow, this is a piece of important information we need to highlight. So I think the warning makes sense. It's just too bad our doc does not auto-generate info/warning section like so docs do.

subhamkrai replied:
we can do something like `> **Warning:**` which we usually do for warning in doc.

sp98 commented:
> @sp98 LGTM. Are you still testing, or it's good to go?

@travisn I couldn't get a cluster with mixed device types to test this. So just manually edited the deviceClass on multi node minikube cluster from `hdd` to `sdd`.  Added an object and confirmed that primary OSD was one with `ssd` deviceclass.  Added details in the PR description. Please let me know if this testing should suffice.

leseb commented on pkg/daemon/ceph/client/pool.go:
```diff
@@ -400,26 +412,50 @@ func createTwoStepCrushRule(context *clusterd.Context, clusterInfo *ClusterInfo,
 
+	crushMap, err := getCurrentCrushMap(context, clusterInfo)
+	if err != nil {
+		return errors.Wrapf(err, "failed create two step crush rule %q", ruleName)
```
```suggestion
		return errors.Wrap(err, "failed get current crush map")
```

leseb commented on pkg/daemon/ceph/client/pool.go:
```diff
@@ -400,26 +412,50 @@ func createTwoStepCrushRule(context *clusterd.Context, clusterInfo *ClusterInfo,
+	crushMap, err := getCurrentCrushMap(context, clusterInfo)
 	if err != nil {
-		return errors.Wrap(err, "failed to get crush map")
+		return errors.Wrapf(err, "failed create hybrid crush rule %q", ruleName)
```
ditto

sp98 commented:
> How about adding to PendingReleaseNotes.md? We're close enough to 1.7 I don't think we need to backport this feature, right?

Updated PendingReleaseNotes.md.  Agree on not backporting this feature.
